client/repo.c

```diff
@@ -636,41 +795,53 @@ TDNFGetRepoMD(
         BAIL_ON_TDNF_ERROR(dwError);
         if(metalink)
         {
-            dwError = TDNFAllocateStringPrintf(
-                          &pszTmpRepoMetalinkFile,
-                          "%s/%s",
-                          pszTmpRepoDataDir,
-                          TDNF_REPO_METALINK_FILE_NAME);
+            dwError = TDNFAllocateStringPrintf(&pszTmpRepoMetalinkFile,
+                                               "%s/%s",
+                                               pszTmpRepoDataDir,
+                                               TDNF_REPO_METALINK_FILE_NAME);
             BAIL_ON_TDNF_ERROR(dwError);
-            dwError = TDNFDownloadFile(
-                          pTdnf,
-                          pRepoData->pszId,
-                          pszRepoMetalink,
-                          pszTmpRepoMetalinkFile,
-                          pRepoData->pszId,
-                          metalink,
-                          &ml_file);
+            dwError = TDNFAllocateStringPrintf(&pszTempBaseUrlFile,
```

Comment:
nDownload flag can be improved by checking the shasum of metalink file and download the repomd and other xml files if the shasum of metalink differs. That way we can avoid the logic of downloading repomd xml and checking the shasum. We can add a separate PR to improvise this logic.